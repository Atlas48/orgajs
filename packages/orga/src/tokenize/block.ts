import { Reader } from '../reader'
import { Token } from '../types'

interface Props {
  reader: Reader;
}

export default ({ reader }: Props) : Token[] => {
  const { match, eat } = reader

  let m = match(/^\s*#\+begin_([^\s]+)\s*(.*)$/i)
  if (m) {
    eat('line')
    const params = m.captures[2].split(' ').map(p => p.trim()).filter(String)
    return [{
      type: 'block.begin',
      params,
      position: m.position,
    }]
  }

  m = match(/^\s*#\+end_([^\s]+)$/i)
  if (m) {
    eat('line')
    return [{
      type: 'block.end',
      position: m.position,
    }]
  }

  return []
}
