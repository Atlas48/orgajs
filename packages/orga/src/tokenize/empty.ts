import { Reader } from 'text-kit'
import { Token } from '../types'

export default ({
  isStartOfLine,
  match,
  endOfLine,
  jump,
  eat,
}: Reader): Token | void => {
  if (!isStartOfLine()) return
  const m = match(/^\s*$/, { end: endOfLine() })
  if (!m) return
  jump(m.position.end)
  return {
    type: 'emptyLine',
    position: m.position,
  }
}
