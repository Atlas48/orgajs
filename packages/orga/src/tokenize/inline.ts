import { Point, Position } from 'unist';
import { after } from '../position';
import { Reader } from '../reader';
import { Token, PhrasingContent, StyledText } from '../../types';
import uri from '../uri';


const LINK_PATTERN = /\[\[([^\]]*)\](?:\[([^\]]*)\])?\]/m; // \1 => link, \2 => text
const FOOTNOTE_PATTERN = /\[fn:(\w+)\]/

const PRE = `[\\s\\({'"]|^`
const POST = `[\\s-\\.,:!?'\\)}]|$`
const BORDER = `[^,'"\\s]`

function markup(marker: string) {
  return RegExp(`(?<=(${PRE}))${marker}(${BORDER}(?:.*?(?:${BORDER}))??)${marker}(?=(${POST}.*))`, 'm')
}

const shift = (point: Point, offset: Point) => {
  return {
    line: point.line + offset.line,
    column: point.column + offset.column,
  }
}

const shiftPosition = (position: Position, offset: Point) => {
  return {
    start: shift(position.start, offset),
    end: shift(position.end, offset),
  }
}

interface Props {
  reader: Reader;
  start?: Point;
  end?: Point;
}



export const tokenize = ({ reader, start, end } : Props): Token[] => {
  const { now, eol, match, jump, substring } = reader
  const s = start || now()
  const e = end || eol()

  let tokens: PhrasingContent[] = [
    {
      type: 'text.plain',
      value: substring({ start: s, end: e }),
      position: { start: s, end: e },
    }
  ]

  const parse = <T extends PhrasingContent>(
    type: T['type'],
    pattern: RegExp,
    content: PhrasingContent[],
    build: (match: { captures: string[], position: Position }) => PhrasingContent,
  ): PhrasingContent[] => {

    return content.reduce((all, token) => {
      if (token.type !== 'text.plain') return all.concat(token)
      const m = match(pattern, token.position)
      if (!m) return all.concat(token)
      if (!token.position || !m.position) {
        throw Error('not gonna happen')
      }
      if (after(token.position.start)(m.position.start)) {
        const position = {
          start: token.position.start,
          end: m.position.start,
        }
        all.push({
          type: 'text.plain',
          value: substring(position),
          position,
        })
      }


      all.push(build(m))

      if (after(m.position.end)(token.position.end)) {
        const rest = parse(type, pattern, [
          {
            type: 'text.plain',
            value: substring({
              start: m.position.end,
              end: token.position.end,
            }),
            position: {
              start: m.position.end,
              end: token.position.end,
            }}
        ], build)
        all = all.concat(rest)
      }

      return all
    }, [] as PhrasingContent[])
  }

  const parseText = (type: StyledText['type'], pattern: RegExp, content: PhrasingContent[]) =>
    parse(type, pattern, content, ({ position, captures }) => ({
      type,
      value: captures[2],
      position,
    }))

  tokens = parse('link', LINK_PATTERN, tokens, ({ position, captures }) => {
    const linkInfo = uri(captures[1])
    return {
      type: 'link',
      description: captures[2],
      ...linkInfo,
      position,
    }
  })

  tokens = parse('footnote.reference', FOOTNOTE_PATTERN, tokens, ({ position, captures }) => ({
    type: 'footnote.reference',
    label: captures[1],
    position,
  }))

  tokens = parseText('text.bold', markup('\\*'), tokens)
  tokens = parseText('text.verbatim', markup('='), tokens)
  tokens = parseText('text.italic', markup('/'), tokens)
  tokens = parseText('text.strikeThrough', markup('\\+'), tokens)
  tokens = parseText('text.underline', markup('_'), tokens)
  tokens = parseText('text.code', markup('~'), tokens)

  jump(e)
  return tokens.filter(t => {
    return t.type !== 'text.plain' || /\S/.test(t.value)
  })
}
