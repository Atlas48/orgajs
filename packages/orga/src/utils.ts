import { Token } from './types'
import { PhrasingContent } from 'orga';

const matchOperatorsRe = /[|\\{}()[\]^$+*?.]/g

export const escape = (str: string): string => {
  return str.replace(matchOperatorsRe, '\\$&')
}

export const isPhrasingContent = (token: Token): token is PhrasingContent => {
  return token.type.startsWith('text.')
    || token.type === 'footnote.reference'
    || token.type === 'link'
}
