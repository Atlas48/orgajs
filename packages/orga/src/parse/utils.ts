import { Node } from 'unist'
import { push } from '../node'
import { Lexer } from '../tokenize'
import { Parent, Token } from '../types'

export default (lexer: Lexer) => {
  const { peek, eat, save, restore } = lexer
  const collect = (stop: (n: Token) => boolean) => (container: Parent): Parent => {
    const token = peek()
    if (!token || stop(token)) return container
    eat()
    push(container)(token)
    return collect(stop)(container)
  }

  const skip = (predicate: (token: Token) => boolean): void => {
    const token = peek()
    if (token && predicate(token)) {
      eat()
      skip(predicate)
      return
    }
    return
  }

  const tryTo = <T extends Node>(
    parse: (lexer: Lexer) => T | undefined,
    action: (node: T) => void,
  ): boolean => {
    const savePoint = save()
    const node = parse(lexer)
    if (!node) {
      restore(savePoint)
      return false
    }
    action(node)
    return true
  }

  return {
    collect,
    skip,
    tryTo,
  }
}
