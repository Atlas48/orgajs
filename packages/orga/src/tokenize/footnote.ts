import { Reader } from '../reader'
import { Token } from '../types'
import { tokenize as tokenizeInline } from './inline'

interface Props {
  reader: Reader;
}

export default ({ reader }: Props) : Token[] => {
  const { match, jump, skipWhitespaces } = reader
  let tokens: Token[] = []
  const m = match(/^\[fn:([^\]]+)\](?=\s)/)
  if (!m) return []
  tokens.push({
    type: 'footnote.label',
    label: m.captures[1],
    position: m.position,
  })
  jump(m.position.end)
  skipWhitespaces()

  tokens = tokens.concat(tokenizeInline({ reader }))

  return tokens
}
