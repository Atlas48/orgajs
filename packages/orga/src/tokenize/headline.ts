import { isEmpty } from '../position'
import { Reader } from '../reader'
import { TodoKeywordSet } from '../todo-keyword-set'
import { Token } from '../types'
import { tokenize } from './inline'

interface Props {
  reader: Reader;
  todoKeywordSets: TodoKeywordSet[];
}

export default ({ reader, todoKeywordSets }: Props) : Token[] => {

  const {
    match,
    skipWhitespaces,
    now,
    eol,
    distance,
    eat,
    jump,
    substring,
  } = reader

  // TODO: cache this, for performance sake
  const todos = todoKeywordSets.flatMap(s => s.keywords)

  const isActionable = (keyword: string): boolean => {
    return !!todoKeywordSets.find(s => s.actionables.includes(keyword))
  }

  let buffer: Token[] = []
  const stars = eat(/^\*+(?=\s)/)
  if (isEmpty(stars)) throw Error('not gonna happen')
  buffer.push({
    type: 'stars',
    level: distance(stars),
    data: { level: distance(stars) },
    position: stars,
  })
  skipWhitespaces()
  const keyword = eat(RegExp(`^${todos.map(escape).join('|')}(?=\\s)`))
  if (!isEmpty(keyword)) {
    const value = substring(keyword)
    buffer.push({
      type: 'todo',
      keyword: value,
      actionable: isActionable(value),
      position: keyword,
    })
  }
  skipWhitespaces()
  const priority = eat(/^\[#(A|B|C)\](?=\s)/)
  if (!isEmpty(priority)) {
    buffer.push({
      type: 'priority',
      value: substring(priority),
      position: priority,
    })
  }

  skipWhitespaces()

  const tags = match(/\s+(:(?:[\w@]+:)+)[ \t]*$/gm)
  let contentEnd = eol()
  if (tags) {
    contentEnd = tags.position.start
  }

  const tokens = tokenize({ reader, end: contentEnd })

  buffer = buffer.concat(tokens)


  if (tags) {
    skipWhitespaces()
    const tagsPosition = { start: now(), end: tags.position.end }
    const s = substring(tagsPosition)
    buffer.push({
      type: 'tags',
      tags: s.split(':'),
      position: { start: now(), end: tags.position.end },
    })
    jump(tags.position.end)
  }
  eat('line')
  return buffer
}
