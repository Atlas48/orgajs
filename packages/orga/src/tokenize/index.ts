import { read } from 'text-kit'
import { Position } from 'unist'
import defaultOptions, { ParseOptions } from '../options'
import todoKeywordSet, { TodoKeywordSet } from '../todo-keyword-set'
import { Token } from '../types'
import tokenizeBlock from './block'
import tokenizeDrawer from './drawer'
import tokenizeFootnote from './footnote'
import tokenizeHeadline from './headline'
import { tokenize as inlineTok } from './inline'
import tokenizeListItem from './list'
import tokenizePlanning from './planning'
import tokenizeTable from './table'

const PLANNING_KEYWORDS = ['DEADLINE', 'SCHEDULED', 'CLOSED']

export interface Lexer {
  eat: (type?: string) => Token | undefined
  eatAll: (type: string) => number
  peek: (offset?: number) => Token | undefined
  match: (cond: RegExp | string, offset?: number) => boolean
  all: () => Token[]
  save: () => number
  restore: (point: number) => void
  addInBufferTodoKeywords: (text: string) => void
  substring: (position: Position) => string
  /** Modify the next token (or the token at the given offset). */
  modify(f: (t: Token) => Token, offset?: number): void
}

export const tokenize = (
  text: string,
  options: Partial<ParseOptions> = {}
): Lexer => {
  const { timezone, todos } = { ...defaultOptions, ...options }

  const reader = read(text)

  const { isStartOfLine, now, eat, getLine, getChar, match, substring } = reader

  const globalTodoKeywordSets = todos.map(todoKeywordSet)

  const inBufferTodoKeywordSets: TodoKeywordSet[] = []

  const todoKeywordSets = () => {
    return inBufferTodoKeywordSets.length === 0
      ? globalTodoKeywordSets
      : inBufferTodoKeywordSets
  }

  let tokens: Token[] = []

  let cursor = 0

  const tok = (): Token[] => {
    eat('whitespaces')
    if (!getChar()) return []

    if (getChar() === '\n') {
      return [
        {
          type: 'newline',
          position: eat('char').position,
        },
      ]
    }

    if (reader.isStartOfLine && match(/^\*+\s+/)) {
      return tokenizeHeadline({
        reader,
        todoKeywordSets: todoKeywordSets(),
      })
    }

    const l = getLine()
    if (PLANNING_KEYWORDS.some((k) => l.startsWith(k))) {
      return tokenizePlanning({
        reader,
        keywords: PLANNING_KEYWORDS,
        timezone,
      })
    }

    if (match(/^#\+/)) {
      const keyword = match(/^\s*#\+(\w+):\s*(.*)$/)
      if (keyword) {
        eat('line')
        return [
          {
            type: 'keyword',
            key: keyword.result[1],
            value: keyword.result[2],
            position: keyword.position,
          },
        ]
      }

      const block = tokenizeBlock(reader)
      if (block.length > 0) return block
    }

    const list = tokenizeListItem(reader)
    if (list.length > 0) return list

    if (match(/^#\s/)) {
      const comment = match(/^#\s+(.*)$/)
      if (comment) {
        eat('line')
        return [
          {
            type: 'comment',
            position: comment.position,
            value: comment.result[1],
          },
        ]
      }
    }

    const drawer = tokenizeDrawer(reader)
    if (drawer.length > 0) return drawer
    const table = tokenizeTable(reader)
    if (table.length > 0) return table

    const hr = eat(/^\s*-{5,}\s*$/)
    if (hr) {
      return [
        {
          type: 'hr',
          position: hr.position,
        },
      ]
    }

    if (now().column === 1) {
      const footnote = tokenizeFootnote(reader)
      if (footnote.length > 0) return footnote
    }

    // last resort
    return inlineTok(reader)
  }

  const peek = (offset = 0): Token | undefined => {
    const pos = cursor + offset
    if (pos >= tokens.length) {
      tokens = tokens.concat(tok())
    }
    return tokens[pos]
  }

  const modify = (f: (t: Token) => Token, offset = 0): void => {
    const pos = cursor + offset
    const token = peek(offset)
    if (token !== undefined) {
      tokens[pos] = f(token)
    }
  }

  const _eat = (type: string | undefined = undefined): Token | undefined => {
    const t = peek()
    if (!t) return undefined
    if (!type || type === t.type) {
      cursor += 1
      return t
    }
    return undefined
  }

  return {
    peek,
    eat: _eat,
    eatAll: (type: string): number => {
      let count = 0
      while (_eat(type)) {
        count += 1
      }
      return count
    },
    match: (cond, offset = 0) => {
      const token = peek()
      if (!token) return false
      if (typeof cond === 'string') {
        return token.type === cond
      }
      return cond.test(token.type)
    },

    all: (max: number | undefined = undefined): Token[] => {
      let _all: Token[] = []
      let tokens = tok()
      while (tokens.length > 0) {
        _all = _all.concat(tokens)
        tokens = tok()
      }
      return _all
    },

    save: () => cursor,

    restore: (point) => {
      cursor = point
    },

    addInBufferTodoKeywords: (text) => {
      inBufferTodoKeywordSets.push(todoKeywordSet(text))
    },
    substring: (pos) => reader.substring(pos.start, pos.end),
    modify,
  }
}
