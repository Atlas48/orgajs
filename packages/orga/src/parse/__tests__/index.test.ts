import fs from 'fs'
import { tokenize } from '../../tokenize'
import { parse } from './utils'
import path from 'path'

it.skip('works', () => {
  expect(
    parse(
      `
#+TITLE: hello world
#+TODO: TODO NEXT | DONE
#+DATE: 2018-01-01

* NEXT headline one
DEADLINE: <2018-01-01 Mon>
:PROPERTIES:
key: value
key: value
:END:

[[https://github.com/xiaoxinghu/orgajs][Here's]] to the *crazy* ones, the /misfits/, the _rebels_, the ~troublemakers~,
the round pegs in the +round+ square holes...
`.trim()
    )
  ).toMatchSnapshot()
})

// describe.skip('parse', () => {
//   const inputs = fs.readdirSync(__dirname)
//     .filter(f => f.endsWith('.org'))

//   test.each(inputs)('could handle %s', async (input) => {

//     const text = await fs.promises.readFile(path.resolve(__dirname, input), 'utf8')

//     const tree = parse(text)
//     // const serilizedTree = JSON.stringify(tree)
//     console.log(`testing ${input}`)
//     console.log({ tree, text })

//   })

// })
