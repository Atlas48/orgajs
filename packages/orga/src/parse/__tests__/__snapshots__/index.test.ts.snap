// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "actionable": true,
          "children": Array [
            Object {
              "level": 1,
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 2,
                  "line": 5,
                  "offset": 67,
                },
                "start": Object {
                  "column": 1,
                  "line": 5,
                  "offset": 66,
                },
              },
              "type": "stars",
            },
            Object {
              "actionable": true,
              "keyword": "NEXT",
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 7,
                  "line": 5,
                  "offset": 72,
                },
                "start": Object {
                  "column": 3,
                  "line": 5,
                  "offset": 68,
                },
              },
              "type": "todo",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 20,
                  "line": 5,
                  "offset": 85,
                },
                "start": Object {
                  "column": 8,
                  "line": 5,
                  "offset": 73,
                },
              },
              "type": "text.plain",
              "value": "headline one",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 1,
                  "line": 6,
                  "offset": 86,
                },
                "start": Object {
                  "column": 20,
                  "line": 5,
                  "offset": 85,
                },
              },
              "type": "newline",
            },
          ],
          "content": "headline one",
          "keyword": "NEXT",
          "level": 1,
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 1,
              "line": 6,
              "offset": 86,
            },
            "start": Object {
              "column": 1,
              "line": 5,
              "offset": 66,
            },
          },
          "type": "headline",
        },
        Object {
          "keyword": "DEADLINE",
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 27,
              "line": 6,
            },
            "start": Object {
              "column": 1,
              "line": 6,
            },
          },
          "timestamp": Object {
            "date": 2017-12-31T11:00:00.000Z,
            "end": undefined,
          },
          "type": "planning",
        },
        Object {
          "name": "PROPERTIES",
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 6,
              "line": 10,
              "offset": 153,
            },
            "start": Object {
              "column": 1,
              "line": 7,
              "offset": 113,
            },
          },
          "type": "drawer",
          "value": "key: value
key: value",
        },
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "description": "Here's",
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 49,
                  "line": 12,
                  "offset": 203,
                },
                "start": Object {
                  "column": 1,
                  "line": 12,
                  "offset": 155,
                },
              },
              "protocol": "https",
              "search": undefined,
              "type": "link",
              "value": "https://github.com/xiaoxinghu/orgajs",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 57,
                  "line": 12,
                  "offset": 211,
                },
                "start": Object {
                  "column": 49,
                  "line": 12,
                  "offset": 203,
                },
              },
              "type": "text.plain",
              "value": " to the ",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 64,
                  "line": 12,
                  "offset": 218,
                },
                "start": Object {
                  "column": 57,
                  "line": 12,
                  "offset": 211,
                },
              },
              "type": "text.bold",
              "value": "crazy",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 75,
                  "line": 12,
                  "offset": 229,
                },
                "start": Object {
                  "column": 64,
                  "line": 12,
                  "offset": 218,
                },
              },
              "type": "text.plain",
              "value": " ones, the ",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 84,
                  "line": 12,
                  "offset": 238,
                },
                "start": Object {
                  "column": 75,
                  "line": 12,
                  "offset": 229,
                },
              },
              "type": "text.italic",
              "value": "misfits",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 90,
                  "line": 12,
                  "offset": 244,
                },
                "start": Object {
                  "column": 84,
                  "line": 12,
                  "offset": 238,
                },
              },
              "type": "text.plain",
              "value": ", the ",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 98,
                  "line": 12,
                  "offset": 252,
                },
                "start": Object {
                  "column": 90,
                  "line": 12,
                  "offset": 244,
                },
              },
              "type": "text.underline",
              "value": "rebels",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 104,
                  "line": 12,
                  "offset": 258,
                },
                "start": Object {
                  "column": 98,
                  "line": 12,
                  "offset": 252,
                },
              },
              "type": "text.plain",
              "value": ", the ",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 119,
                  "line": 12,
                  "offset": 273,
                },
                "start": Object {
                  "column": 104,
                  "line": 12,
                  "offset": 258,
                },
              },
              "type": "text.code",
              "value": "troublemakers",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 120,
                  "line": 12,
                  "offset": 274,
                },
                "start": Object {
                  "column": 119,
                  "line": 12,
                  "offset": 273,
                },
              },
              "type": "text.plain",
              "value": ",",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 1,
                  "line": 13,
                  "offset": 275,
                },
                "start": Object {
                  "column": 120,
                  "line": 12,
                  "offset": 274,
                },
              },
              "type": "text.plain",
              "value": " ",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 23,
                  "line": 13,
                  "offset": 297,
                },
                "start": Object {
                  "column": 1,
                  "line": 13,
                  "offset": 275,
                },
              },
              "type": "text.plain",
              "value": "the round pegs in the ",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 30,
                  "line": 13,
                  "offset": 304,
                },
                "start": Object {
                  "column": 23,
                  "line": 13,
                  "offset": 297,
                },
              },
              "type": "text.strikeThrough",
              "value": "round",
            },
            Object {
              "parent": [Circular],
              "position": Object {
                "end": Object {
                  "column": 46,
                  "line": 13,
                  "offset": 320,
                },
                "start": Object {
                  "column": 30,
                  "line": 13,
                  "offset": 304,
                },
              },
              "type": "text.plain",
              "value": " square holes...",
            },
          ],
          "parent": [Circular],
          "position": Object {
            "end": Object {
              "column": 46,
              "line": 13,
              "offset": 320,
            },
            "start": Object {
              "column": 1,
              "line": 12,
              "offset": 155,
            },
          },
          "type": "paragraph",
        },
      ],
      "level": 1,
      "parent": [Circular],
      "position": Object {
        "end": Object {
          "column": 46,
          "line": 13,
          "offset": 320,
        },
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 66,
        },
      },
      "properties": Object {
        "date": "2018-01-01",
        "title": "hello world",
      },
      "type": "section",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 46,
      "line": 13,
      "offset": 320,
    },
    "start": Object {
      "column": 1,
      "line": 5,
      "offset": 66,
    },
  },
  "properties": Object {
    "date": "2018-01-01",
    "title": "hello world",
  },
  "type": "document",
}
`;
